### PR TITLE
[api][spinel] add support for logging crash dumps

### PR DESCRIPTION
**Pull request by lmnotran**
This PR adds a new feature that allows platforms to log crash logs.

### Additions
- `void otPlatLogCrashDump(void)` - API that logs a crash dump using OpenThread Logging APIs
- `SPINEL_PROP_RCP_LOG_CRASH_DUMP` - spinel prop that calls `otPlatLogCrashDump()` when `Set`
- `SPINEL_CAP_RCP_LOG_CRASH_DUMP` - spinel capability denoting that a RCP supports logging crash dumps

### Usage
- For `ot-cli-ftd|mtd`, `otPlatLogCrashDump()` is called at the end of app initialization. See [main.c]
- For Host/RCP setups, during initialization, the Host gets the RCP capabilities. If the RCP has the `SPINEL_CAP_RCP_LOG_CRASH_DUMP` capability, the Host will `Set` the `SPINEL_PROP_RCP_LOG_CRASH_DUMP` property, triggering the RCP to call `otPlatLogCrashDump()`.

### CODE REVIEW

**Comment by zhanglongxia** [comment]
The main purpose of this commit is to make the RCP to upload the crash log after it crashes. Here are my understandings (please correct me if I'm wrong):
<1> RCP crashes, it stores the crash info to Flash or non-init RAM and then it restarts.
<2> RCP sends the [RESET_REASON](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L934) to the host when it starts.
<3> OpenThread host restarts after it [receives the RESET_REASION](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel.cpp#L669).
<4> OpenThread host sends `SPINEL_PROP_RCP_LOG_CRASH_DUMP ` to the RCP after it detects that the RCP support uploading the crash info when it starts.
<5> RCP sends the crash info to OpenThread host via `SPINEL_PROP_RCP_LOG_CRASH_DUMP ` after it receives the `SPINEL_PROP_RCP_LOG_CRASH_DUMP`.


I think there is no need to change any OpenThread source code, you can directly call `otLogCritPlat()` in [ExtensionBase::SignalNcpInit()](https://github.com/openthread/openthread/blob/main/src/core/instance/extension_example.cpp#L79) to upload the crash info on the RCP side in your private extension source code. Then the process is like this:

<1> RCP crashes, it stores the crash info to Flash or non-init RAM and then it restarts.
<2> RCP sends the crash info to the OpenThread host, then sending the [RESET_REASON](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L934) to the host.
<3> OpenThread host receives the crash log and writes the log to the log system.
<4> OpenThread host restarts after it [receives the RESET_REASION](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel.cpp#L669).

**Comment by lmnotran** [comment]
@jwhui could you add reviewers, please?

File: include/openthread/platform/misc.h [hunk 1]
```diff
@@ -207,6 +207,14 @@ otError otPlatSetMcuPowerState(otInstance *aInstance, otPlatMcuPowerState aState
  */
 otPlatMcuPowerState otPlatGetMcuPowerState(otInstance *aInstance);
 
+/**
+ * Logs a crash dump using OpenThread logging APIs
+ *
+ * @retval OT_ERROR_NONE            Crash dump was logged successfully
+ * @retval OT_ERROR_NOT_CAPABLE     Platform is not capable of logging a crash dump
+ */
+otError otPlatLogCrashDump(void);
```
Comment (abtink) [inline include/openthread/platform/misc.h:216]: Does adding this as a `otPlat` API make sense?
- `otPlat` APIs are intended for OT core modules use.
-  We don't use this from any OT API.
-  It is used in `ncp_base.cpp` and from `main` (CLI)

**Comment by abtink** [comment]
I agree with @zhanglongxia  suggestions in https://github.com/openthread/openthread/pull/9829#issuecomment-1925563823

**Comment by lmnotran** [comment]
> The main purpose of this commit is to make the RCP to upload the crash log after it crashes. Here are my understandings (please correct me if I'm wrong): <1> RCP crashes, it stores the crash info to Flash or non-init RAM and then it restarts. <2> RCP sends the [RESET_REASON](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L934) to the host when it starts. <3> OpenThread host restarts after it [receives the RESET_REASION](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel.cpp#L669). <4> OpenThread host sends `SPINEL_PROP_RCP_LOG_CRASH_DUMP ` to the RCP after it detects that the RCP support uploading the crash info when it starts. <5> RCP sends the crash info to OpenThread host via `SPINEL_PROP_RCP_LOG_CRASH_DUMP ` after it receives the `SPINEL_PROP_RCP_LOG_CRASH_DUMP`.
> 
> I think there is no need to change any OpenThread source code, you can directly call `otLogCritPlat()` in [ExtensionBase::SignalNcpInit()](https://github.com/openthread/openthread/blob/main/src/core/instance/extension_example.cpp#L79) to upload the crash info on the RCP side in your private extension source code. Then the process is like this:
> 
> <1> RCP crashes, it stores the crash info to Flash or non-init RAM and then it restarts. <2> RCP sends the crash info to the OpenThread host, then sending the [RESET_REASON](https://github.com/openthread/openthread/blob/main/src/ncp/ncp_base.cpp#L934) to the host. <3> OpenThread host receives the crash log and writes the log to the log system. <4> OpenThread host restarts after it [receives the RESET_REASION](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel.cpp#L669).

@zhanglongxia Thanks for the suggestion! I never thought to use the vendor hooks to implement this. I tried out your suggestion and was able to capture the crash dump in `ot-cli`'s log. 
[ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt](https://github.com/openthread/openthread/files/14302670/ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt)

While this does work without modifying the stack, it does have the limitation of logs only being printed once. If the Host misses the logs for any reason, there's no way for the Host to query the RCP for a crash dump. Do you think it this ability is something that would be valuable to add to the stack?

**Comment by zhanglongxia** [comment]
> @zhanglongxia Thanks for the suggestion! I never thought to use the vendor hooks to implement this. I tried out your suggestion and was able to capture the crash dump in `ot-cli`'s log. [ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt](https://github.com/openthread/openthread/files/14302670/ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt)
> 
> While this does work without modifying the stack, it does have the limitation of logs only being printed once. If the Host misses the logs for any reason, there's no way for the Host to query the RCP for a crash dump. Do you think it this ability is something that would be valuable to add to the stack?

It depends on how RCP handles crash log. For example, RCP could keep the crash log and send the crash log to host when [RCP calls the `otPlatRadioGetIeeeEui64()`](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel.cpp#L161) and then RCP clears the crash log. Or RCP could provide a diag command for the host to retrieve the crash log. So I think it's not necessary to add additional spinel properties to process the crash log.

**Comment by lmnotran** [comment]
Those are good points. Thanks for the reviews @zhanglongxia @abtink !

**Pull request closed without merging**